mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraAttachmentDAO.java

```diff
@@ -0,0 +1,119 @@
+import static org.apache.james.mailbox.cassandra.table.CassandraAttachmentTable.FIELDS;
+import static org.apache.james.mailbox.cassandra.table.CassandraAttachmentTable.ID;
+import static org.apache.james.mailbox.cassandra.table.CassandraAttachmentTable.PAYLOAD;
+import static org.apache.james.mailbox.cassandra.table.CassandraAttachmentTable.SIZE;
+import static org.apache.james.mailbox.cassandra.table.CassandraAttachmentTable.TABLE_NAME;
+import static org.apache.james.mailbox.cassandra.table.CassandraAttachmentTable.TYPE;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+import javax.inject.Inject;
+
+import org.apache.james.backends.cassandra.utils.CassandraAsyncExecutor;
+import org.apache.james.mailbox.model.Attachment;
+import org.apache.james.mailbox.model.AttachmentId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.Row;
+import com.datastax.driver.core.Session;
+import com.google.common.base.Preconditions;
+
+public class CassandraAttachmentDAO {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CassandraAttachmentMapper.class);
+    private static final boolean NO_LOG_IF_EMPTY = false;
+
+    private final CassandraAsyncExecutor cassandraAsyncExecutor;
+    private final PreparedStatement insertStatement;
+    private final PreparedStatement selectStatement;
+
+    @Inject
+    public CassandraAttachmentDAO(Session session) {
+        this.cassandraAsyncExecutor = new CassandraAsyncExecutor(session);
+
+        this.selectStatement = prepareSelect(session);
+        this.insertStatement = prepareInsert(session);
+    }
+
+    private PreparedStatement prepareInsert(Session session) {
+        return session.prepare(
+            insertInto(TABLE_NAME)
+                .value(ID, bindMarker(ID))
+                .value(PAYLOAD, bindMarker(PAYLOAD))
+                .value(TYPE, bindMarker(TYPE))
+                .value(SIZE, bindMarker(SIZE)));
+    }
+
+    private PreparedStatement prepareSelect(Session session) {
+        return session.prepare(select(FIELDS)
+            .from(TABLE_NAME)
+            .where(eq(ID, bindMarker(ID))));
+    }
+
+    public CompletableFuture<Optional<Attachment>> getAttachment(AttachmentId attachmentId) {
+        return getAttachment(attachmentId, NO_LOG_IF_EMPTY);
+    }
+
+    public CompletableFuture<Optional<Attachment>> getAttachment(AttachmentId attachmentId, boolean logIfEmpty) {
+        Preconditions.checkArgument(attachmentId != null);
+        return cassandraAsyncExecutor.executeSingleRow(
+            selectStatement.bind()
+                .setString(ID, attachmentId.getId()))
+            .thenApply(optional -> optional.map(this::attachment))
+            .thenApply(optional -> logNotFound(attachmentId, logIfEmpty, optional));
```

Comment:
What you propose? I've been looking for the equivalement of Stream::peek but can not find it...

I believe a way around would be:

```
CompletableFutureUtil.peek(future, Consummer<T>)
```

But this would hurt readability a lot IMO. I found current approach better.